dali/kernels/signal/wavelet/wavelet_gpu.cu

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/kernels/signal/wavelet/wavelet_gpu.cuh"
+#include <cmath>
+#include <complex>
+#include <vector>
+#include "dali/core/common.h"
+#include "dali/core/error_handling.h"
+#include "dali/core/format.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/signal/wavelet/mother_wavelet.cuh"
+#include "dali/core/tensor_shape.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+
+// computes wavelet value for each sample in specified range,
+// and each a and b coeff
+template <typename T, template <typename> class W >
+__global__ void ComputeWavelet(const SampleDesc<T>* sample_data, W<T> wavelet) {
+  // id inside block
+  const int64_t b_id = threadIdx.y * blockDim.x + threadIdx.x;
+  // wavelet sample id
+  const int64_t t_id = blockDim.x * blockDim.y * blockIdx.x + b_id;
+  auto& sample = sample_data[blockIdx.z];
+  if (t_id >= sample.size_in) return;
+  __shared__ T shm[1025];
+  auto a = sample.a[blockIdx.y];
+  auto x = std::pow(2.0, a);
+  if (a == 0.0) {
+    shm[b_id] = sample.in[t_id];
+  }
+  else {
+    shm[b_id] = x * sample.in[t_id];
+    shm[1024] = std::pow(2.0, a / 2.0);
+  }
+  for (int i = 0; i < sample.size_b; ++i) {
```

Comment:
Shouldn't you sync here to make sure that all writes to shm are done?

Reply:
Right. I've added sync across threads.